streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java

```
@@ -196,69 +207,47 @@ public long streamTime() {
         return streamTimeSupplier.get();
     }
 
-    private abstract static class StateStoreReadOnlyDecorator<T extends StateStore> implements StateStore {
+    private abstract static class StateStoreReadOnlyDecorator<T extends StateStore> extends AbstractStateStore {
         static final String ERROR_MESSAGE = "Global store is read only";
 
-        final T underlying;
-
-        StateStoreReadOnlyDecorator(final T underlying) {
-            this.underlying = underlying;
-        }
-
-        @Override
-        public String name() {
-            return underlying.name();
+        StateStoreReadOnlyDecorator(final StateStore inner) {
+            super(inner);
         }
 
-        @Override
-        public void init(final ProcessorContext context, final StateStore root) {
```

GlobalStores should also be guarded with this PR, ie, `init()` and `close()` for read-only-decorator should throw an exception, too.